staging/src/k8s.io/client-go/tools/cache/reflector.go

```diff
@@ -335,6 +338,11 @@ func (r *Reflector) ListAndWatch(stopCh <-chan struct{}) error {
 					klog.V(2).Infof("%s: watch of %v returned 429 - backing off", r.name, r.expectedTypeName)
 					<-r.initConnBackoffManager.Backoff().C()
 					continue
+				case apierrors.IsInternalError(err):
```

Comment:
I'm somewhat reluctant to retrying every single internal error. Can we try to somehow distinguish the no-leader error that we really care about here from a generic internal error?

There are many other potential reasons for internal-errors that I don't think retrying here is a good idea in a generic case (the typical usecase of reflector, not the one that we have for watchcache).

Reply:
I would prefer not to hard-code "etcdserver: no leader" error as if it changes this mitigation would stop working. In worst case, we will terminate watchers 30 seconds later if we get other non-retriable error here in case of watchcache.

Also, I don't think this should be used by reflectors other than watchcache most likely - by default this mitigation is turned off - MaxInternalErrorRetryDuration is set to 0.

Reply:
Hi @wojtek-t , can you please explain more what Internal error should be retried with a Watch and what internal error should be treated more seriously by a List?

for instance, should the below short buffer error be followed by a Watch or a List?

**watch of *v1.Service ended with: an error on the server ("unable to decode an event from the watch stream: got short buffer with n=0, base=4092, cap=163840")**

Reply:
You can see above for the example - no-leader is the best example.

The one you pointed out I've never seen and I'm not sure what this "short buffer" really means - sounds like a problem with allocating more memory... ?

Reply:
So if it was the case you wrote above, then retrying with Watch is the good thing. Basically, whenever retrying with Watch has a high chance of a success, it's desired, fallback to LIST should be the last resort.

Reply:
@wojtek-t 
just would like to know, why from the start of design of `Reflector` we are treating watch error with a List?
Early code in `cache/reflector.go` will only retry a watch with a watch if apierrors.IsTooManyRequests occurs

```
	for {
......
		if w == nil {
			timeoutSeconds := int64(minWatchTimeout.Seconds() * (rand.Float64() + 1.0))
			options := metav1.ListOptions{
......
			}

			w, err = r.listerWatcher.Watch(options)
			if err != nil {
				if shouldContinue := handleInitialWatchRequestError(r.initConnBackoffManager, err); shouldContinue {
					continue
				}
				return err
			}
		}

		_, err = watchHandler(r, start, w, r.store, r.expectedType, r.expectedGVK, r.name, r.expectedTypeName, r.setLastSyncResourceVersion, r.clock, false, resyncerrc, stopCh)
		w.Stop()
		w = nil // always create a new watch request
		if err != nil {
			if err != errorStopRequested {
				switch {
				case isExpiredError(err):
					// Don't set LastSyncResourceVersionUnavailable - LIST call with ResourceVersion=RV already
					// has a semantic that it returns data at least as fresh as provided RV.
					// So first try to LIST with setting RV to resource version of last observed object.
					klog.V(4).Infof("%s: watch of %v closed with: %v", r.name, r.expectedTypeName, err)
				case apierrors.IsTooManyRequests(err):
					klog.V(2).Infof("%s: watch of %v returned 429 - backing off", r.name, r.expectedTypeName)
					<-r.initConnBackoffManager.Backoff().C()
					continue
				default:
					klog.Warningf("%s: watch of %v ended with: %v", r.name, r.expectedTypeName, err)
				}
			}
			return nil
		}
	}
```